src/runtime/security/kerberos_utils.cpp

```diff
@@ -0,0 +1,313 @@
+    if (!err.is_ok()) {
+        dwarn_f("get credentials of {} from KDC failed, reason({})",
+                _principal_name,
+                err.description());
+        return err;
+    }
+    auto cleanup = dsn::defer([&]() { krb5_free_cred_contents(g_krb5_context, &creds); });
+
+    // store credentials into _ccache.
+    WRAP_KRB5_ERR(krb5_cc_store_cred(g_krb5_context, _ccache, &creds), err, "store_cred");
+    if (!err.is_ok()) {
+        dwarn_f(
+            "store credentials of {} to cache failed, err({})", _principal_name, err.description());
+        return err;
+    }
+
+    _cred_expire_timestamp = creds.times.endtime;
+    ddebug_f("get credentials of {} from KDC ok, expires at {}",
+             _principal_name,
+             utils::time_s_to_date_time(_cred_expire_timestamp));
+    return err;
+}
+
+error_s kinit_context::get_formatted_identities()
+{
+    char *tmp_str = nullptr;
+    KRB5_RETURN_NOT_OK(krb5_unparse_name(g_krb5_context, _principal, &tmp_str),
+                       "unparse principal name failed");
+    auto cleanup = dsn::defer([&]() { krb5_free_unparsed_name(g_krb5_context, tmp_str); });
+    _principal_name = tmp_str;
+
+    return parse_username_from_principal(_principal, _username);
+}
+
+error_s kinit_context::kinit(const std::string &keytab_file, const std::string &principal)
+{
+    // create a krb5 library context.
+    init_krb5_ctx();
+
+    // convert a string principal name to a krb5_principal structure.
+    KRB5_RETURN_NOT_OK(krb5_parse_name(g_krb5_context, principal.c_str(), &_principal),
+                       "couldn't parse principal");
+
+    // get _principal_name and _user_name from _principal
+    RETURN_NOT_OK(get_formatted_identities());
+
+    // get a handle for a key table.
+    KRB5_RETURN_NOT_OK(krb5_kt_resolve(g_krb5_context, keytab_file.c_str(), &_keytab),
+                       "couldn't resolve keytab file");
+
+    // acquire credential cache handle
+    KRB5_RETURN_NOT_OK(krb5_cc_default(g_krb5_context, &_ccache),
+                       "couldn't acquire credential cache handle");
+
+    // initialize credential cache
+    KRB5_RETURN_NOT_OK(krb5_cc_initialize(g_krb5_context, _ccache, _principal),
+                       "initialize credential cache failed");
+
+    // allocate a new initial credential options structure
+    KRB5_RETURN_NOT_OK(krb5_get_init_creds_opt_alloc(g_krb5_context, &_opt),
+                       "alloc get_init_creds_opt structure failed");
+
+    // get and schedule to renew credentials from KDC and store it into _ccache
+    RETURN_NOT_OK(get_credentials());
+    schedule_renew_credentials();
+
+    ddebug_f("logged in from keytab as {}, local username {}", _principal_name, _username);
+    return error_s::make(ERR_OK);
+}
+
+#undef KRB5_RETURN_NOT_OK // only used in this internal namespace
+#undef WRAP_KRB5_ERR      // only used in this internal namespace
+} // namespace internal
+
+error_s check_configuration()
+{
```

Comment:
Add `dassert(FLAGS_enable_auth)` here to ensure and tell reviewers this functions only executed when FLAGS_enable_auth is true.